File: Java/PancakeSort.java
public class PancakeSort {
    public static void main(String[] args) {
        int[] arr = {3, 6, 2, 7, 5, 8, 4, 1};
        pancakeSort(arr);
        printArray(arr);
    }

    public static void pancakeSort(int[] arr) {
        for (int currSize = arr.length; currSize > 1; --currSize) {
            int maxIndex = findMax(arr, currSize);
            if (maxIndex != currSize - 1) {
                flip(arr, maxIndex);
                flip(arr, currSize - 1);
            }
        }
    }

    private static int findMax(int[] arr, int n) {
        int maxIndex = 0;
        for (int i = 1; i < n; i++) {
            if (arr[i] > arr[maxIndex]) {
                maxIndex = i;
            }
        }
        return maxIndex;
    }

    private static void flip(int[] arr, int end) {
        int start = 0;
        while (start < end) {
            int temp = arr[start];
            arr[start] = arr[end];
            arr[end] = temp;
            start++;
            end--;
        }
    }

    private static void printArray(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            if (i > 0) {
                System.out.print(", ");
            }
            System.out.print(arr[i]);
        }
        System.out.println();
    }
}

File: Kotlin/pancakeSort.kt
fun flip(arr: IntArray, k: Int) {
    var start = 0
    var end = k
    while (start < end) {
        val temp = arr[start]
        arr[start] = arr[end]
        arr[end] = temp
        start++
        end--
    }
}

fun findMax(arr: IntArray, n: Int): Int {
    var maxIndex = 0
    for (i in 1 until n) {
        if (arr[i] > arr[maxIndex]) {
            maxIndex = i
        }
    }
    return maxIndex
}

fun pancakeSort(arr: IntArray) {
    for (currSize in arr.size downTo 2) {
        val maxIndex = findMax(arr, currSize)
        if (maxIndex != currSize - 1) {
            flip(arr, maxIndex)
            flip(arr, currSize - 1)
        }
    }
}

fun main() {
    val arr = intArrayOf(3, 6, 1, 10, 7, 2, 0, -2, 4)
    pancakeSort(arr)
    println(arr.joinToString(", "))
}